### PR TITLE
[touchwand] Touchwand Binding initial contribution  - migration to OH3 

### PR DESCRIPTION
**Pull request by roieg**
TouchWand Wanderfull controller binding initial contribution

[touchwand][WIP] initial contribution
This is OH3 PR migration from [PR #6310 ](https://github.com/openhab/openhab-addons/pull/6310)

Touchwand Wanderfull™ Hub basic is a plug & play Z-Wave based controller that uses Wi-Fi and Bluetooth to easily connect all smart home components.
The binding allow to connect to Touchwand Wanderfull hub and control shutters switches dimmers and wallcontrollers
TouchWand [website](http://www.touchwand.com)
This is initial contribution , currently the binding supports

- Touchwand Hub auto discovery
- Units auto discovery 
- Switches
- Shutters
- Dimmers

### CODE REVIEW

**Review by fwolter — APPROVED** [review]
As I already approved the old PR, I will do so here, trusting you, that you didn't change anything significantly in regard to the old PR.

**Comment by roieg** [comment]
> As I already approved the old PR, I will do so here, trusting you, that you didn't change anything significantly in regard to the old PR.

@fwolter Thank you for your review and effort.
Indeed the code is the same as in PR #6310 , added only the minor changes needed for OH3.
This include almost all of the changes requested by @Hilbrand in his reviews (beside one I could not fully understand).

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:68]: why is this synchronized?

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:57]: This wan't do anything in the constructor, you need to move this to the `activate` method instead.

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:142]:
```suggestion
    protected void stopScan() {
```

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/dto/Csc.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/dto/Csc.java:21]: Can you give this class a better name?

File: bundles/org.openhab.binding.touchwand/src/main/resources/OH-INF/thing/wallcontroller.xml
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/resources/OH-INF/thing/wallcontroller.xml:19]: `wallcontroller` isn't a valid `item-type`

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java:113]: If `dispose` is called immediately after `initialize` your binding should be able to shutdown without side effects.
In this case it is possible for `dispose` to be called before you have initialized the `touchWandWebSockets` field.
To prevent this from happening you need use a boolean `isRunning` field along with synchronization to make sure that access to the `isRunning` and touchWandWebSockets field is atomic.

You could synchronize on a new lock object, but for this fix it is less messy to just synchronize on the `this` object instead.
```
private volatile boolean isRunning = false;

public synchronized void initialize() {
```

```suggestion
        isRunning = true;
        scheduler.execute(() -> {
            boolean thingReachable = false;
            String password = config.password;
            String username = config.username;
            thingReachable = touchWandClient.connect(username, password, host, port.toString());
            if (thingReachable) {
                updateStatus(ThingStatus.ONLINE);
                synchronized(this){
                    if(isRunning){
                        TouchWandWebSockets localSockets = touchWandWebSockets = new TouchWandWebSockets(host, scheduler);
                        localSockets.registerListener(this);
                        localSockets.connect();
                    }
                }
            } else {
                updateStatus(ThingStatus.OFFLINE);
            }
        });
    }

    @Override
    public void handleCommand(ChannelUID channelUID, Command command) {
    }

    public boolean isAddSecondaryControllerUnits() {
        return addSecondaryUnits;
    }

    public int getStatusRefreshTime() {
        return statusRefreshRateSec;
    }

    @Override
    public synchronized void dispose() {
        isRunning = false;
        TouchWandWebSockets myTouchWandWebSockets = touchWandWebSockets;
        if (myTouchWandWebSockets != null) {
            myTouchWandWebSockets.unregisterListener(this);
            myTouchWandWebSockets.dispose();
        }
    }
```

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:53]: We can take advantage of java11 here. yay!
```suggestion
    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Set.of(THING_TYPE_SHUTTER, THING_TYPE_SWITCH, THING_TYPE_WALLCONTROLLER, THING_TYPE_DIMMER);
```

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java:55]: please move these fields to below your static constants

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:68]: package private?

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:105]: Shouldn't the `socketReconnect` be canceled here?

**Review by cpmeister — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:199]: done

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java]: done - aligned both to new Date().getTime()

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java]: done

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/dto/Csc.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/dto/Csc.java:21]: I agree it is a bad name .
The class name is from the Json exposed to me from TouchWand controller and its name is to match the Json field .
This is out of my control , I prefer to keep it this way.

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java]: moved to activate

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java]: done

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java]: done

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:104]: indeed it should....
owe you a beer....

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java:101]: done

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java:113]: done

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:126]: You can remove this

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:105]: Please try to catch specific exceptions here. At the very least catch RuntimeException instead.

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:118]:
```suggestion
        private DatagramSocket mySocket;
```

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:111]:
```suggestion
        localSocketReceivedThread.setName(TouchWandBindingConstants.DISCOVERY_THREAD_ID);
        localSocketReceivedThread.setDaemon(true);
        localSocketReceivedThread.start();
```

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:55]:
```suggestion
    protected @Nullable TouchWandBridgeHandler bridgeHandler;
```

**Review by cpmeister — CHANGES_REQUESTED** [review]
These should be my final change requests. Other than that you look good to go!

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:104]: done

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:126]: done

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:105]: The IDE insists I will catch Exception , catching just RuntimeException  will not do 
Any idea ?

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:55]: done

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java
Comment (roieg) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:111]: done

File: bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java
Comment (cpmeister) [reply bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:105]: Ah, I forgot that `client.stop()` just throws a plain `Exception`. In that case please disregard my recommendation here.

**Review by cpmeister — APPROVED** [review]
LGTM. Just waiting for jenkins to finish before I can merge this.

**Comment by roieg** [comment]
Thank you @fwolter @cpmeister @Hilbrand for your effort  and time spend leveling up this binding

**Pull request merged**
